### PR TITLE
[ES-1292925] Fix metrics with reusable counter resets

### PR DESCRIPTION
**Pull request by jnyi**
This PR tries to fix when a reusable counter metrics resets:

image (1)

* Added a number of unit tested and make sure they pass when doing counter functions.

* Also unit tested if not doing counter functions the original time series is returned

Pending integration tests, sent for early review and feedbacks

Integration tests work as expected:

Screenshot 2024-11-25 at 9 40 56 AM

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: pkg/dedup/merge_iter.go [hunk 1]
```diff
@@ -49,26 +56,25 @@ func (m *mergedSeries) Iterator(_ chunkenc.Iterator) chunkenc.Iterator {
...
+		currentValue: v,
 		cnt:          1,
 	}
 }
 
 // Return true if this is the new majority value.
 func (q *quorumValuePicker) addValue(v float64) bool {
-	iv := int64(v)
-	if q.currentValue == iv {
+	if q.currentValue == v {
```
Comment (hczhu-db) [inline pkg/dedup/merge_iter.go:72]: Using ` == ` to compare float values is usually not recommended, but I think it should be fine in this context because the values come from the same source of truth.

File: pkg/dedup/merge_iter.go [hunk 2]
```diff
@@ -12,30 +12,37 @@ import (
...
+	iters := make([]adjustableSeriesIterator, 0, len(m.replicas))
 	oks := make([]bool, 0, len(m.replicas))
 	for _, r := range m.replicas {
-		it := r.Iterator(nil)
+		var it adjustableSeriesIterator
+		if m.isCounter {
+			it = &counterErrAdjustSeriesIterator{Iterator: r.Iterator(nil)}
+		} else {
+			it = &noopAdjustableSeriesIterator{Iterator: r.Iterator(nil)}
+		}
```
Comment (hczhu-db) [inline pkg/dedup/merge_iter.go:45]: You are applying counter adjustments to the raw data samples before deduplication. This is more complicated than applying the adjustments to the single time series after deduplication. The adjustments will intervene with quorum-based deduplication logic. I'm concerned it may introduce other edge cases.

Comment (hczhu-db) [reply pkg/dedup/merge_iter.go:45]: Can you add another test case such that one raw time series has a large gap with a reset, but the other two have complete data?
```
replica 0: [[1000, 10], [10000, 8], [11000, 10]]

replica 1: [[1000, 10], [2000, 0], [3000, 1], [4000, 2], [5000, 3], [6000, 4], [7000, 5], [8000, 6], [9000, 7],  [10000, 8], [11000, 10]
replica 2: [[1000, 10], [2000, 0], [3000, 1], [4000, 2], [5000, 3], [6000, 4], [7000, 5], [8000, 6], [9000, 7],  [10000, 8], [11000, 10]
```

Comment (jnyi) [reply pkg/dedup/merge_iter.go:45]: added new test cases, the reason i've done it this way is to reuse `counterErrAdjustSeriesIterator` which you need to call `adjustAtValue()` somewhere, passing a merged time series to original newDedupSeries doesn't work

Comment (jnyi) [reply pkg/dedup/merge_iter.go:45]: added the test you suggested above

File: pkg/dedup/iter.go [hunk 3]
```diff
@@ -160,9 +167,13 @@ func (s *dedupSeriesSet) At() storage.Series {
...
 	repl := make([]storage.Series, len(s.replicas))
 	copy(repl, s.replicas)
-	if s.f == UseMergedSeries {
+	if s.deduplicationFunc == AlgorithmQuorum {
 		// merge all samples which are ingested via receiver, no skips.
-		return NewMergedSeries(s.lset, repl)
+		// feed the merged series into dedup series which apply counter adjustment
+		return NewMergedSeries(s.lset, repl, s.f)
+	}
+	if s.deduplicationFunc == AlgorithmChain {
```
Comment (jnyi) [inline pkg/dedup/iter.go:174]: fyi @yuchen-db, I've also added the prometheus implementation, but it doesn't work for a number of unit tests if you wonder

File: pkg/dedup/merge_iter.go (hunk 1 above)
Comment (jnyi) [reply pkg/dedup/merge_iter.go:72]: agreed, the value should be the same for given timestamp

File: pkg/dedup/iter.go [hunk 4]
```diff
@@ -23,6 +23,12 @@ import (
 // that timestamps are in milliseconds and sampling frequencies typically multiple seconds long.
 const initialPenalty = 5000
 
+const (
```
Comment (davidyuanfs) [inline pkg/dedup/iter.go:26]: Is there a wiki or comment I can tell these three dedup different for handling series?

Comment (jnyi) [reply pkg/dedup/iter.go:26]: it is in command flag, the penalty one is original thanos OSS implementation, the second one is prometheus, quorum is what we added to fix a number of issues we see during dedup.

Comment (jnyi) [reply pkg/dedup/iter.go:26]: I feel there are naunces to tell in words, mostly it is covered by unit tests as it deals with numbers

File: pkg/query/querier.go [hunk 5]
```diff
@@ -210,18 +210,20 @@ func newQuerierWithOpts(
 
 	partialResponseStrategy := storepb.PartialResponseStrategy_ABORT
 	if opts.GroupReplicaPartialResponseStrategy {
-		level.Debug(logger).Log("msg", "Enabled group-replica partial response strategy in newQuerierInternal")
+		level.Info(logger).Log("msg", "Enabled group-replica partial response strategy in newQuerierInternal")
```
Comment (hczhu-db) [inline pkg/query/querier.go]: This will be very chatty. I intentionally changed it to Debug previously.

**Review by hczhu-db — CHANGES_REQUESTED** [review]

File: pkg/query/querier.go [hunk 6]
```diff
@@ -210,18 +210,20 @@ func newQuerierWithOpts(
...
 		partialResponseStrategy = storepb.PartialResponseStrategy_GROUP_REPLICA
 	} else if partialResponse {
 		partialResponseStrategy = storepb.PartialResponseStrategy_WARN
 	}
+	level.Info(logger).Log("msg", "Deduplication algorithm applied", "func", opts.DeduplicationFunc)
```
Comment (hczhu-db) [inline pkg/query/querier.go]: Every query will print this log.

File: pkg/query/querier.go (hunk 5 above)
Comment (hczhu-db) [reply pkg/query/querier.go]: Every query will print this log.

**Review by hczhu-db — CHANGES_REQUESTED** [review]

File: pkg/dedup/iter.go [hunk 7]
```diff
@@ -23,6 +23,12 @@ import (
...
+	AlgorithmPenalty = "penalty"
```
Comment (davidyuanfs) [inline pkg/dedup/iter.go:27]: These three dedup will change the alert behaviors, how can we validate in the future to avoid regression when updating this?

File: pkg/query/querier.go (hunk 5 above)
Comment (jnyi) [reply pkg/query/querier.go]: found i actually doesn't need this log line, it is logged here already: 

```
level.Info(logger).Log("msg", "databricks querier features", "opts", fmt.Sprintf("%+v", opts))
```

File: pkg/dedup/iter.go (hunk 7 above)
Comment (jnyi) [reply pkg/dedup/iter.go:27]: we should use `quorum` all the time, people shouldn't make change and need to go through shard flags and rigorous testing

**Review by hczhu-db — APPROVED** [review]
